File: scripts/check_sources.py
#!/usr/bin/env python3
"""
Prüft alle Quellen aus data/source-registry.json auf Änderungen und schreibt
das Ergebnis nach data/source-status.json (von faq.html per fetch() geladen).

Prüfmethoden pro Quelle (Feld "check.method" in der Registry):
  ckan       - opendata.muenchen.de CKAN-API: vergleicht das Änderungsdatum mit dem
               beim letzten Lauf gespeicherten. Nutzt bevorzugt das Extra-Feld
               "GeoPortal Last Modified Date" (echtes Geodaten-Änderungsdatum bei
               Geoportal-Datensätzen), sonst den jüngsten Resource-Zeitstempel.
               CKANs metadata_modified wird bewusst NICHT genutzt: bei Geoportal-
               Datensätzen wird täglich neu geharvestet (neue Resource-IDs) auch
               ohne echte Datenänderung.
  heuristic  - keine Live-Quelle vorhanden (RIS-Mitgliederdaten ändern sich nur bei
               Nachrücken/Rücktritt, nicht zeitbasiert prüfbar). Nutzt das Datum der
               letzten lokalen Änderung (git log) + den bekannten Turnus - ist die
               Datei länger als der Turnus nicht angefasst worden, gilt eine Prüfung
               als fällig.

Aufruf: python3 scripts/check_sources.py [--dry-run]
"""

import json
import subprocess
import sys
import urllib.request
import ssl
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
REGISTRY_PATH = ROOT / "data" / "source-registry.json"
STATUS_PATH = ROOT / "data" / "source-status.json"

TURNUS_DAYS = {
    "jährlich": 365,
}


def now_iso():
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def git_last_changed(rel_path):
    try:
        out = subprocess.run(
            ["git", "log", "-1", "--format=%ad", "--date=short", "--", rel_path],
            cwd=ROOT, capture_output=True, text=True, timeout=10,
        )
        d = out.stdout.strip()
        return d or None
    except Exception:
        return None


def fetch(url, timeout=30):
    ctx = ssl.create_default_context()
    ctx.load_verify_locations("/etc/ssl/cert.pem")
    req = urllib.request.Request(url, headers={"User-Agent": "rathausmuc-source-check/1.0"})
    with urllib.request.urlopen(req, context=ctx, timeout=timeout) as resp:
        return resp.read()


def check_ckan(dataset, prev):
    url = f"https://opendata.muenchen.de/api/3/action/package_show?id={dataset}"
    try:
        data = json.loads(fetch(url))
        if not data.get("success"):
            raise RuntimeError(data.get("error"))
        result = data["result"]
        extras = {e["key"]: e["value"] for e in result.get("extras", [])}

        if "GeoPortal Last Modified Date" in extras:
            source_modified = extras["GeoPortal Last Modified Date"]
            source_field = "GeoPortal Last Modified Date"
        else:
            resources = result.get("resources", [])
            stamps = [r.get("last_modified") or r.get("created") for r in resources]
            stamps = [s for s in stamps if s]
            source_modified = max(stamps) if stamps else result.get("metadata_modified")
            source_field = "resource last_modified/created"
    except Exception as e:
        return {"status": "unknown", "detail": f"CKAN-Abruf fehlgeschlagen: {e}"}

    prev_modified = (prev or {}).get("source_modified")
    changed = prev_modified is not None and prev_modified != source_modified
    return {
        "status": "changed" if changed else "ok",
        "source_modified": source_modified,
        "detail": (f"Quelle wurde seit letzter Prüfung aktualisiert (laut {source_field})." if changed
                   else f"Laut Portal zuletzt geändert am {source_modified} ({source_field}), keine Änderung seit letzter Prüfung."),
    }


def check_heuristic(source):
    files = source.get("files", [])
    dates = [git_last_changed(f) for f in files]
    dates = [d for d in dates if d]
    last_changed = max(dates) if dates else None

    turnus = source.get("turnus", "")
    interval_days = None
    for key, days in TURNUS_DAYS.items():
        if key in turnus:
            interval_days = days
            break

    if not last_changed or interval_days is None:
        return {
            "status": "manual",
            "local_updated": last_changed,
            "detail": "Kein automatischer Änderungscheck möglich (keine Live-API, kein fester Turnus). Turnus: " + (turnus or "unbekannt"),
        }

    age_days = (datetime.now().date() - datetime.strptime(last_changed, "%Y-%m-%d").date()).days
    due = age_days > interval_days
    return {
        "status": "check_due" if due else "ok",
        "local_updated": last_changed,
        "detail": (f"Seit {age_days} Tagen nicht aktualisiert, Turnus ist '{turnus}' – Prüfung beim Original empfohlen." if due
                   else f"Zuletzt aktualisiert vor {age_days} Tagen, Turnus '{turnus}' noch nicht erreicht."),
    }


def main():
    dry_run = "--dry-run" in sys.argv
    registry = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
    prev_status = {}
    if STATUS_PATH.exists():
        prev_status = json.loads(STATUS_PATH.read_text(encoding="utf-8")).get("sources", {})

    results = {}
    for cat in registry["categories"]:
        for source in cat["sources"]:
            sid = source["id"]
            method = source["check"]["method"]
            local_updated = None
            for f in source.get("files", []):
                d = git_last_changed(f)
                if d:
                    local_updated = d
                    break

            if method == "ckan":
                result = check_ckan(source["check"]["dataset"], prev_status.get(sid))
            elif method == "heuristic":
                result = check_heuristic(source)
            else:
                result = {"status": "unknown", "detail": f"Unbekannte Methode: {method}"}

            result.setdefault("local_updated", local_updated)
            result["method"] = method
            result["checked_at"] = now_iso()
            results[sid] = result
            print(f"{sid:30s} [{method:10s}] -> {result['status']}")

    if dry_run:
        print("\n--dry-run: nichts geschrieben.")
        return

    STATUS_PATH.write_text(
        json.dumps({"generated_at": now_iso(), "sources": results}, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    print(f"\nGeschrieben: {STATUS_PATH.relative_to(ROOT)}")


if __name__ == "__main__":
    main()

File: cron/check_sources.php
<?php
/**
 * Quellen-Status-Check
 *
 * Cronjob-URL: https://rathausmuc.de/cron/check_sources.php?token=DEIN_CRON_TOKEN
 * Intervall: wöchentlich (KAS-Cronjob)
 *
 * PHP-Port von scripts/check_sources.py - läuft serverseitig statt lokal, weil auf dem
 * All-Inkl-Hosting kein Python zur Verfügung steht. Die Logik ist bewusst zweigeteilt:
 *
 *   1. data/source-status.json wird bei JEDEM Lauf neu berechnet und automatisch direkt
 *      nach main committet (rein informative Ampel auf faq.html, ändert keine
 *      Anzeigedaten, daher unkritisch für Auto-Commit).
 *   2. Erkennt der Check bei einer "ckan"-Quelle eine echte Änderung gegenüber dem letzten
 *      bekannten Stand, wird zusätzlich ein GitHub Issue angelegt ("Quelle X hat sich
 *      geändert: alt -> neu"). Es wird NICHT versucht, die betroffene CSV automatisch neu
 *      zu berechnen - die CSV-Aktualisierung bleibt ein manueller Schritt.
 *
 * "heuristic"-Quellen (stadtrat.csv, ba01-25.csv - keine Live-API, ändern sich nur bei
 * Nachrücken/Rücktritt) werden ohne Netzwerk-Call geprüft: lokales Änderungsdatum (aus dem
 * letzten Stand übernommen, da hier kein Git-Zugriff besteht) gegen den bekannten Turnus.
 * Das lokale Datum wird nur dann aktuell gehalten, wenn jemand lokal
 * scripts/check_sources.py laufen lässt (das hat Git-Zugriff) und committet.
 */

require __DIR__ . '/_cfg.php';
require __DIR__ . '/github_api.php';

// Token-Check
$token = $_GET['token'] ?? '';
if (!$token || $token !== $cfg['cron']['token']) {
    http_response_code(403);
    echo json_encode(['error' => 'forbidden']);
    exit;
}

$githubToken = $cfg['github']['token'];
$githubRepo  = $cfg['github']['repo'];
$now         = date('c');
$debug       = isset($_GET['debug']);

try {
    $registryRaw = file_get_contents(__DIR__ . '/../data/source-registry.json');
    if ($registryRaw === false) {
        throw new RuntimeException('data/source-registry.json nicht lesbar');
    }
    $registry = json_decode($registryRaw, true);
    if (!$registry) {
        throw new RuntimeException('data/source-registry.json ist kein gültiges JSON');
    }

    $prevStatusRaw = @file_get_contents(__DIR__ . '/../data/source-status.json');
    $prevStatus    = $prevStatusRaw ? (json_decode($prevStatusRaw, true)['sources'] ?? []) : [];

    $results  = [];
    $changed  = [];

    foreach ($registry['categories'] as $cat) {
        foreach ($cat['sources'] as $src) {
            $id     = $src['id'];
            $method = $src['check']['method'];
            $prev   = $prevStatus[$id] ?? [];

            if ($method === 'ckan') {
                $result = check_ckan($src['check']['dataset'], $prev);
                if ($result['status'] === 'changed') {
                    $changed[] = ['id' => $id, 'title' => $src['title'], 'result' => $result, 'prev' => $prev];
                }
            } elseif ($method === 'heuristic') {
                $result = check_heuristic($src, $prev);
            } else {
                $result = ['status' => 'unknown', 'detail' => "Unbekannte Prüfmethode: {$method}"];
            }

            $result['method']        = $method;
            $result['checked_at']    = $now;
            $result['local_updated'] = $result['local_updated'] ?? ($prev['local_updated'] ?? null);
            $results[$id]            = $result;
        }
    }

    $newStatusJson = json_encode(
        ['generated_at' => $now, 'sources' => $results],
        JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES
    ) . "\n";

    if ($debug) {
        header('Content-Type: application/json');
        echo $newStatusJson;
        exit;
    }

    $statusUnchanged = $prevStatusRaw !== false && normalize_json($prevStatusRaw) === normalize_json($newStatusJson);

    if (!$statusUnchanged) {
        // Direkt auf Festplatte schreiben - sofort sichtbar ohne Deploy-Umweg
        file_put_contents(__DIR__ . '/../data/source-status.json', $newStatusJson);

        github_commit_multiple(
            ['data/source-status.json' => $newStatusJson],
            "data: Quellen-Status aktualisiert ({$now})",
            $githubToken,
            $githubRepo
        );
    }

    $issuesCreated = [];
    foreach ($changed as $c) {
        $issue = github_create_issue(
            "Quelle geändert: {$c['title']}",
            issue_body($c),
            $githubToken,
            $githubRepo
        );
        $issuesCreated[] = $issue['number'] ?? null;
    }

    http_response_code(200);
    echo json_encode([
        'ok'             => true,
        'timestamp'      => $now,
        'checked'        => count($results),
        'status_changed' => !$statusUnchanged,
        'changed'        => count($changed),
        'issues_created' => $issuesCreated,
    ]);

} catch (Exception $e) {
    http_response_code(500);
    echo json_encode(['error' => $e->getMessage()]);
}

function normalize_json(string $json): ?string
{
    $data = json_decode($json, true);
    if ($data === null) {
        return null;
    }
    unset($data['generated_at']);
    $sources = $data['sources'] ?? [];
    foreach ($sources as &$s) {
        unset($s['checked_at']);
    }
    unset($s);
    $data['sources'] = $sources;
    return json_encode($data);
}

function issue_body(array $c): string
{
    $r   = $c['result'];
    $old = $c['prev']['source_modified'] ?? '(unbekannt, erster Check)';
    $new = $r['source_modified'] ?? '(unbekannt)';
    return "Automatischer Wochen-Check hat eine Änderung an der Quelle **{$c['title']}** erkannt.\n\n"
         . "- Zuletzt bekannt: `{$old}`\n"
         . "- Jetzt laut Portal: `{$new}`\n\n"
         . "Bitte in der [Datenquellen-Übersicht](https://rathausmuc.de/faq.html) gegenprüfen "
         . "und die betroffene CSV bei Bedarf manuell aktualisieren. Dieses Issue danach schließen.\n\n"
         . "_Automatisch erstellt von `cron/check_sources.php`._";
}

function check_ckan(string $dataset, array $prev): array
{
    $url = 'https://opendata.muenchen.de/api/3/action/package_show?id=' . urlencode($dataset);
    $ctx = stream_context_create([
        'http' => ['timeout' => 20, 'header' => "User-Agent: rathausmuc-source-check\r\n"],
    ]);
    $raw = @file_get_contents($url, false, $ctx);
    if ($raw === false) {
        return ['status' => 'unknown', 'detail' => 'CKAN-Abruf fehlgeschlagen (keine Antwort).'];
    }
    $data = json_decode($raw, true);
    if (!($data['success'] ?? false)) {
        $msg = $data['error']['message'] ?? 'unbekannter Fehler';
        return ['status' => 'unknown', 'detail' => "CKAN-Abruf fehlgeschlagen: {$msg}"];
    }
    $result = $data['result'];
    $extras = [];
    foreach ($result['extras'] ?? [] as $e) {
        $extras[$e['key']] = $e['value'];
    }

    if (isset($extras['GeoPortal Last Modified Date'])) {
        $sourceModified = $extras['GeoPortal Last Modified Date'];
        $field          = 'GeoPortal Last Modified Date';
    } else {
        $stamps = [];
        foreach ($result['resources'] ?? [] as $res) {
            $stamps[] = $res['last_modified'] ?? $res['created'] ?? null;
        }
        $stamps         = array_values(array_filter($stamps));
        $sourceModified = $stamps ? max($stamps) : ($result['metadata_modified'] ?? null);
        $field          = 'resource last_modified/created';
    }

    $prevModified = $prev['source_modified'] ?? null;
    $isChanged    = $prevModified !== null && $prevModified !== $sourceModified;

    return [
        'status'          => $isChanged ? 'changed' : 'ok',
        'source_modified' => $sourceModified,
        'detail'          => $isChanged
            ? "Quelle wurde seit letzter Prüfung aktualisiert (laut {$field})."
            : "Laut Portal zuletzt geändert am {$sourceModified} ({$field}), keine Änderung seit letzter Prüfung.",
    ];
}

function check_heuristic(array $src, array $prev): array
{
    $localUpdated = $prev['local_updated'] ?? null;
    $turnus       = $src['turnus'] ?? '';

    $intervalDays = null;
    if (str_contains($turnus, 'jährlich')) {
        $intervalDays = 365;
    }

    if (!$localUpdated || $intervalDays === null) {
        return [
            'status'        => 'manual',
            'local_updated' => $localUpdated,
            'detail'        => 'Kein automatischer Änderungscheck möglich (keine Live-API, kein fester Turnus). Turnus: ' . ($turnus ?: 'unbekannt'),
        ];
    }

    $ageDays = (int) floor((strtotime('today') - strtotime($localUpdated)) / 86400);
    $due     = $ageDays > $intervalDays;

    return [
        'status'        => $due ? 'check_due' : 'ok',
        'local_updated' => $localUpdated,
        'detail'        => $due
            ? "Seit {$ageDays} Tagen nicht aktualisiert, Turnus ist '{$turnus}' – Prüfung beim Original empfohlen."
            : "Zuletzt aktualisiert vor {$ageDays} Tagen, Turnus '{$turnus}' noch nicht erreicht.",
    ];
}
